numba/core/typing/new_builtins.py

```diff
@@ -0,0 +1,1178 @@
+            if real == imag == types.np_float32:
+                return signature(types.complex64, real, imag)
+            else:
+                return signature(types.complex128, real, imag)
+
+
+#------------------------------------------------------------------------------
+
+@infer_global(enumerate)
+class Enumerate(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+        it = args[0]
+        if len(args) > 1 and not isinstance(args[1], types.Integer):
+            raise errors.NumbaTypeError("Only integers supported as start "
+                                        "value in enumerate")
+        elif len(args) > 2:
+            #let python raise its own error
+            enumerate(*args)
+
+        if isinstance(it, types.IterableType):
+            enumerate_type = types.EnumerateType(it)
+            return signature(enumerate_type, *args)
+
+
+@infer_global(zip)
+class Zip(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+        if all(isinstance(it, types.IterableType) for it in args):
+            zip_type = types.ZipType(args)
+            return signature(zip_type, *args)
+
+
+@infer_global(iter)
+class Iter(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+        if len(args) == 1:
+            it = args[0]
+            if isinstance(it, types.IterableType):
+                return signature(it.iterator_type, *args)
+
+
+@infer_global(next)
+class Next(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+        if len(args) == 1:
+            it = args[0]
+            if isinstance(it, types.IteratorType):
+                return signature(it.yield_type, *args)
+
+
+#------------------------------------------------------------------------------
+
+@infer_global(type)
+class TypeBuiltin(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+        if len(args) == 1:
+            # One-argument type() -> return the __class__
+            # Avoid literal types
+            arg = types.unliteral(args[0])
+            classty = self.context.resolve_getattr(arg, "__class__")
+            if classty is not None:
+                return signature(classty, *args)
+
+
+#------------------------------------------------------------------------------
+
+@infer_getattr
+class OptionalAttribute(AttributeTemplate):
+    key = types.Optional
+
+    def generic_resolve(self, optional, attr):
+        return self.context.resolve_getattr(optional.type, attr)
+
+#------------------------------------------------------------------------------
+
+@infer_getattr
+class DeferredAttribute(AttributeTemplate):
+    key = types.DeferredType
+
+    def generic_resolve(self, deferred, attr):
+        return self.context.resolve_getattr(deferred.get(), attr)
+
+#------------------------------------------------------------------------------
+
```

Comment:
Note to self: `MinValInfer` for `get_type_min_value` and `get_type_max_value` used to be here.